torch/distributed/device_mesh.py

```diff
@@ -80,9 +80,12 @@ def create_child_mesh(
             # swap the current dim to the last dim then reshape to flatten out other
             # dims, so we can just extract the list of ranks which contains cur_rank.
             cur_rank = device_mesh.get_rank()
-            pg_ranks_by_dim = device_mesh.mesh.swapdims(-1, mesh_dim).reshape(
-                -1, device_mesh.mesh.size(mesh_dim)
+            pg_ranks_by_dim = (
+                device_mesh.mesh.swapdims(-1, mesh_dim)
+                .reshape(-1, device_mesh.mesh.size(mesh_dim))
+                .to(torch.int32)
```

Comment:
Need to convert this mesh to torch.int32. Otherwise, the mesh created by DeviceMesh has a mesh tensor of dtype torch.int32 and mesh sliced out from a parent mesh would have a dtype of torch.int64.

Reply:
Hmm curious why that's the case? The transpose and reshape operations shouldn't change dtype?

Reply:
Somehow it did. When I print out a mesh sliced out from a parent mesh, the dtype becomes torch.int64 instead of torch.int32.

Reply:
Hmm I wrote a simple test
<img width="1019" alt="image" src="https://github.com/pytorch/pytorch/assets/1372575/15bb1354-14f7-4177-8d7f-ecf7a24d8cbc">
<img width="926" alt="image" src="https://github.com/pytorch/pytorch/assets/1372575/7cf8134b-99b1-40fa-9712-d4cb85a0f4c9">

Reply:
Thanks for checking. Seems I could not repro the inconsistency now. After a rebase, I am seeing:
```
mesh_2d=DeviceMesh([[0, 1, 2, 3], [4, 5, 6, 7]], mesh_dim_names=('DP', 'TP')), mesh_2d.mesh.dtype=torch.int64
tp_mesh=DeviceMesh([4, 5, 6, 7], mesh_dim_names=('TP',)), tp_mesh.mesh.dtype=torch.int64
```

Reply:
Ah. Ok. Here is what it is. If I create a mesh thru `init_device_mesh`, then torch.int64. If we do it thru `DeviceMesh` directly, then dtype is torch.int32.